cuebot/src/main/java/com/imageworks/spcue/dispatcher/HostReportHandler.java

```diff
@@ -456,103 +465,266 @@ private void changeLockState(DispatchHost host, CoreDetail coreInfo) {
     }
 
     /**
-     * Handle memory reservations for the given host.  This will re-balance memory
-     * reservations on the machine and kill and frames that are out of control.
-     *
+     * Prevent host from entering an OOM state where oom-killer might start killing important OS processes.
+     * The kill logic will kick in one of the following conditions is met:
+     *   - Host has less than OOM_MEMORY_LEFT_THRESHOLD_PERCENT memory available
+     *   - A frame is taking more than OOM_FRAME_OVERBOARD_PERCENT of what it had reserved
+     * For frames that are using more than they had reserved but not above the threshold, negotiate expanding
+     * the reservations with other frames on the same host
      * @param host
      * @param report
      */
-    private void handleMemoryReservations(final DispatchHost host, final HostReport report) {
+    private void handleMemoryUsage(final DispatchHost host, final HostReport report) {
+        final double OOM_MAX_SAFE_USED_MEMORY_THRESHOLD =
+                env.getRequiredProperty("dispatcher.oom_max_safe_used_memory_threshold", Double.class);
+        final double OOM_FRAME_OVERBOARD_ALLOWED_THRESHOLD =
+                env.getRequiredProperty("dispatcher.oom_frame_overboard_allowed_threshold", Double.class);
+        RenderHost renderHost = report.getHost();
+        List<RunningFrameInfo> runningFrames = report.getFramesList();
+
+        boolean memoryWarning = renderHost.getTotalMem() > 0 &&
+                ((double)renderHost.getFreeMem()/renderHost.getTotalMem() <
+                        (1.0 - OOM_MAX_SAFE_USED_MEMORY_THRESHOLD));
+
+        if (memoryWarning) {
```

Comment:
We might want to add some logging here to indicate what's happening and what the calculate values are. Seems like it could be hard to decipher why a frame is being killed.

Reply:
Logging here would be redundant as the underlying logic implemented on killWorstMemoryOffender already logs for both success and error cases.

Reply:
Ok, I'll leave it up to you. It would be good to log enough detail to know why a frame was killed without having to guess at the logic.